nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java

```diff
@@ -0,0 +1,338 @@
+        innerRelationshipsSet.add(REL_MESSAGE);
+        relationships = Collections.unmodifiableSet(innerRelationshipsSet);
+    }
+
+    @Override
+    public void onPropertyModified(PropertyDescriptor descriptor, String oldValue, String newValue) {
+        // resize the receive buffer, but preserve data
+        if (descriptor == PROP_MAX_QUEUE_SIZE) {
+            // it's a mandatory integer, never null
+            int newSize = Integer.valueOf(newValue);
+            if (mqttQueue != null) {
+                int msgPending = mqttQueue.size();
+                if (msgPending > newSize) {
+                    logger.warn("New receive buffer size ({}) is smaller than the number of messages pending ({}), ignoring resize request. Processor will be invalid.",
+                            new Object[]{newSize, msgPending});
+                    return;
+                }
+                LinkedBlockingQueue<MQTTQueueMessage> newBuffer = new LinkedBlockingQueue<>(newSize);
+                mqttQueue.drainTo(newBuffer);
+                mqttQueue = newBuffer;
+            }
+
+        }
+    }
+
+    @Override
+    public Collection<ValidationResult> customValidate(ValidationContext context) {
+        final Collection<ValidationResult> results = super.customValidate(context);
+        int newSize = context.getProperty(PROP_MAX_QUEUE_SIZE).asInteger();
+        if (mqttQueue == null) {
+            mqttQueue = new LinkedBlockingQueue<>(context.getProperty(PROP_MAX_QUEUE_SIZE).asInteger());
+        }
+        int msgPending = mqttQueue.size();
+        if (msgPending > newSize) {
+            results.add(new ValidationResult.Builder()
+                    .valid(false)
+                    .subject("ConsumeMQTT Configuration")
+                    .explanation(String.format("%s (%d) is smaller than the number of messages pending (%d).",
+                            PROP_MAX_QUEUE_SIZE.getDisplayName(), newSize, msgPending))
+                    .build());
+        }
+
+        return results;
+    }
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        logger = getLogger();
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) throws IOException, ClassNotFoundException {
+        qos = context.getProperty(PROP_QOS).asInteger();
+        maxQueueSize = context.getProperty(PROP_MAX_QUEUE_SIZE).asLong();
+        topicFilter = context.getProperty(PROP_TOPIC_FILTER).getValue();
+
+        buildClient(context);
+        scheduled.set(true);
+    }
+
+    @OnUnscheduled
+    public void onUnscheduled(final ProcessContext context) {
+        scheduled.set(false);
+
+        try {
+            synchronized (mqttClientConnectLock) {
```

Comment:
May be I am missing something, but what's the point of synchronization here? Is there a contention of some type? I mean knowing when/how @OnUnscheduled is invoked i don't seem to see the benefit.